core/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -146,27 +185,12 @@ boolean shardCanBeDeleted(ClusterState state, IndexShardRoutingTable indexShardR
 
         for (ShardRouting shardRouting : indexShardRoutingTable) {
             // be conservative here, check on started, not even active
-            if (!shardRouting.started()) {
+            if (shardRouting.started() == false) {
                 return false;
             }
 
-            // if the allocated or relocation node id doesn't exists in the cluster state  it may be a stale node,
-            // make sure we don't do anything with this until the routing table has properly been rerouted to reflect
-            // the fact that the node does not exists
-            DiscoveryNode node = state.nodes().get(shardRouting.currentNodeId());
```

Comment:
yes, I was as well thinking about an assertion. Putting it on the ClusterState constructor however means that this check will be done a lot of times. We often just make simple changes on cluster states like increasing version or setting local node which could slow down our tests. We have similar checks currently in `RoutingTable.validate` but this is only called after changes made by AllocationService. If we could ensure that DiscoveryNodes and RoutingTable/RoutingNodes are only changed in AllocationService (something which I think is feasible and on my stack of things to investigate), a check in that method would be enough. I'll investigate where to put the assertion (it's complementary to this PR).